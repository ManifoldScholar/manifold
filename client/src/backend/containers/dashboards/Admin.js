import React, { useMemo } from "react";
import EntitiesList, {
  Button,
  Search,
  ProjectRow,
  JournalRow
} from "backend/components/list/EntitiesList";
import lh from "helpers/linkHandler";
import { useFetch, usePaginationState, useAuthentication } from "hooks";
import { projectsAPI, journalsAPI } from "api";
import withFilteredLists, {
  projectFilters,
  journalFilters
} from "hoc/withFilteredLists";
import Authorize from "hoc/Authorize";
import DashboardComponents from "backend/components/dashboard";
import PageHeader from "backend/components/layout/PageHeader";
import Authorization from "helpers/authorization";
import isEqual from "lodash/isEqual";
import { useTranslation } from "react-i18next";

export function DashboardsAdminContainer({
  entitiesListSearchProps,
  entitiesListSearchParams
}) {
  const [projectsPagination, setProjectsPageNumber] = usePaginationState(1, 5);
  const [journalsPagination, setJournalsPageNumber] = usePaginationState(1, 5);

  const authentication = useAuthentication();

  const projectFiltersWithDefaults = useMemo(
    () => ({
      withUpdateAbility: true,
      ...entitiesListSearchParams.projects
    }),
    [entitiesListSearchParams?.projects]
  );

  const journalFiltersWithDefaults = useMemo(
    () => ({
      withUpdateOrIssueUpdateAbility: true,
      ...entitiesListSearchParams.journals
    }),
    [entitiesListSearchParams?.journals]
  );

  const { data: projects, meta: projectsMeta } = useFetch({
    request: [projectsAPI.index, projectFiltersWithDefaults, projectsPagination]
  });

  const { data: journals, meta: journalsMeta } = useFetch({
    request: [journalsAPI.index, journalFiltersWithDefaults, journalsPagination]
  });

  const authorization = new Authorization();
  const canCreateProjects = authorization.authorizeAbility({
    authentication,
    entity: "project",
    ability: "create"
  });

  const { t } = useTranslation();

  const noProjects =
    isEqual(
      entitiesListSearchParams?.projects,
      entitiesListSearchParams?.initialProjects
    ) && canCreateProjects
      ? t("dashboard.empty_message_creator")
      : t("dashboard.empty_message_generic");

  if (!projects) return null;

  const projectList =
    projects && projectsMeta ? (
      <div className="dashboard-panel">
        <div className="panel">
          <EntitiesList
            entities={projects}
            entityComponent={ProjectRow}
            entityComponentProps={{
              placeholderMode: "small"
            }}
            title={t("glossary.project_title_case", {
              count: projectsMeta?.pagination?.totalCount
            })}
            titleLink={lh.link("backendProjects")}
            titleIcon="BEProject64"
            titleStyle="bar"
            titleTag="h2"
            showCount
            showCountInTitle
            unit={t("glossary.project", {
              count: projectsMeta?.pagination?.totalCount
            })}
            pagination={projectsMeta.pagination}
            paginationPadding={1}
            callbacks={{
              onPageClick: page => () => setProjectsPageNumber(page)
            }}
            emptyMessage={noProjects}
            search={
              <Search
                searchStyle="vertical"
                onFilterChange={() => setProjectsPageNumber(1)}
                {...entitiesListSearchProps("projects")}
              />
            }
            buttons={[
              <Button
                path={lh.link("backendProjectsNew")}
                text={t("projects.add_button_label")}
                authorizedFor="project"
                authorizedTo="create"
                type="add"
              />
            ]}
          />
        </div>
      </div>
    ) : null;

  const journalList =
    journals && journalsMeta ? (
      <div className="dashboard-panel">
        <div className="panel">
          <EntitiesList
            entities={journals}
            entityComponent={JournalRow}
            entityComponentProps={{
              placeholderMode: "small"
            }}
            title={t("glossary.journal_title_case", {
              count: journalsMeta?.pagination?.totalCount
            })}
            titleLink={lh.link("backendJournals")}
            titleIcon="Journals64"
            titleStyle="bar"
            titleTag="h2"
            showCount
            showCountInTitle
            unit={t("glossary.journal", {
              count: journalsMeta?.pagination?.totalCount
            })}
            pagination={journalsMeta.pagination}
            paginationPadding={1}
            callbacks={{
              onPageClick: page => () => setJournalsPageNumber(page)
            }}
            search={
              <Search
                searchStyle="vertical"
                onFilterChange={() => setJournalsPageNumber(1)}
                {...entitiesListSearchProps("journals")}
              />
            }
            buttons={[
              <Button
                path={lh.link("backendJournalsNew")}
                text={t("journals.add_button_label")}
                authorizedFor="journal"
                authorizedTo="create"
                type="add"
              />
            ]}
          />
        </div>
      </div>
    ) : null;

  const analytics = (
    <Authorize entity="statistics" ability={"read"}>
      <PageHeader
        type="analytics"
        link={{
          path: lh.link("backendAnalytics"),
          label: t("actions.see_all")
        }}
        titleTag="h2"
        title={t("analytics.global_header")}
      />

      <DashboardComponents.Analytics />
    </Authorize>
  );

  const canSeeAnalytics = authorization.authorizeAbility({
    authentication,
    entity: "statistics",
    ability: "read"
  });

  const guts = canSeeAnalytics ? (
    <>
      <div className="left">
        {projectList}
        {journalList}
      </div>
      <div className="right">{analytics}</div>
    </>
  ) : (
    <>
      <div className="left">{projectList}</div>
      <div className="right">{journalList}</div>
    </>
  );

  return (
    <main id="skip-to-main" tabIndex={-1}>
      <h1 className="screen-reader-text">{t("dashboard.title")}</h1>
      <section>
        <div className="container">
          <section className="backend-dashboard">{guts}</section>
        </div>
      </section>
    </main>
  );
}
export default withFilteredLists(DashboardsAdminContainer, {
  projects: projectFilters({ snapshotState: true }),
  journals: journalFilters({ snapshotState: true })
});
