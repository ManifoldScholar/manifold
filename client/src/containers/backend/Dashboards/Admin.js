import React, { PureComponent } from "react";
import PropTypes from "prop-types";
import connectAndFetch from "utils/connectAndFetch";
import { entityStoreActions } from "actions";
import {
  List,
  Project,
  Dashboard as DashboardComponents
} from "components/backend";
import { HigherOrder } from "containers/global";
import { select, meta } from "utils/entityUtils";
import { projectsAPI, statisticsAPI, requests } from "api";
import debounce from "lodash/debounce";
import Authorization from "helpers/authorization";

const { request } = entityStoreActions;

const perPage = 5;

export class DashboardsAdminContainer extends PureComponent {
  static mapStateToProps = state => {
    return {
      statistics: select(requests.beStats, state.entityStore),
      projects: select(requests.beProjects, state.entityStore),
      projectsMeta: meta(requests.beProjects, state.entityStore),
      recentProjects: select(requests.beRecentProjects, state.entityStore),
      authentication: state.authentication,
      projectsListSnapshot:
        state.ui.transitory.stateSnapshots.dashboardProjectsList
    };
  };

  static propTypes = {
    projects: PropTypes.array,
    statistics: PropTypes.object,
    dispatch: PropTypes.func,
    projectsMeta: PropTypes.object,
    recentProjects: PropTypes.array,
    authentication: PropTypes.object,
    projectsListSnapshot: PropTypes.object.isRequired,
    snapshotCreator: PropTypes.func.isRequired
  };

  constructor(props) {
    super(props);
    this.state = this.initialState(props);
    this.authorization = new Authorization();
    this.updateResults = debounce(this.updateResults.bind(this), 250);
  }

  componentDidMount() {
    const projectsRequest = request(
      projectsAPI.index(this.buildFetchFilter(this.props, this.state.filter), {
        number: this.props.projectsListSnapshot.page,
        size: perPage
      }),
      requests.beProjects
    );
    const recentProjectsRequest = request(
      projectsAPI.index(
        this.buildFetchFilter(this.props, { order: "updated_at DESC" }),
        { size: 2 }
      ),
      requests.beRecentProjects
    );
    const statsRequest = request(statisticsAPI.show(), requests.beStats);

    const { promise: one } = this.props.dispatch(projectsRequest);
    const { promise: two } = this.props.dispatch(recentProjectsRequest);
    const promises = [one, two];

    const readStats = this.authorization.authorizeAbility({
      authentication: this.props.authentication,
      entity: "statistics",
      ability: "read"
    });
    if (readStats) {
      const { promise: three } = this.props.dispatch(statsRequest);
      promises.push(three);
    }

    return Promise.all(promises);
  }

  initialState(props) {
    return Object.assign({}, { filter: props.projectsListSnapshot.filter });
  }

  buildFetchFilter = (props, base) => {
    const out = Object.assign({}, base);
    const currentUser = props.authentication.currentUser;
    if (!currentUser) return out;
    if (currentUser.attributes.abilities.viewDrafts) return out;
    out.withUpdateAbility = true;
    return out;
  };

  snapshotState(page) {
    const snapshot = { filter: this.state.filter, page };
    this.props.snapshotCreator(snapshot);
  }

  updateResults(eventIgnored = null, page = 1) {
    this.snapshotState(page);

    const pagination = { number: page, size: perPage };
    const action = request(
      projectsAPI.index(
        this.buildFetchFilter(this.props, this.state.filter),
        pagination
      ),
      requests.beProjects
    );
    this.props.dispatch(action);
  }

  filterChangeHandler = filter => {
    this.setState({ filter }, () => {
      this.updateResults();
    });
  };

  updateHandlerCreator = page => {
    return event => {
      this.updateResults(event, page);
    };
  };

  renderProjectCount = () => {
    if (!this.props.projectsMeta) return null;

    const { totalCount } = this.props.projectsMeta.pagination;
    const label = totalCount > 1 || totalCount === 0 ? "projects" : "project";

    return <span className="list-total">{`${totalCount} ${label}`}</span>;
  };

  render() {
    return (
      <div>
        <section>
          <div className="container">
            <section className="backend-dashboard">
              <div className="left">
                <header className="section-heading-secondary">
                  <h3>
                    <i className="manicon manicon-stack" aria-hidden="true" />
                    {"Projects"} {this.renderProjectCount()}
                  </h3>
                </header>
                {this.props.projects && this.props.projectsMeta ? (
                  <List.Searchable
                    showEntityCount={false}
                    initialFilter={this.state.filter}
                    defaultFilter={{ order: "sort_title ASC" }}
                    entities={this.props.projects}
                    pagination={this.props.projectsMeta.pagination}
                    paginationClickHandler={this.updateHandlerCreator}
                    paginationClass="secondary"
                    entityComponent={Project.ListItem}
                    filterChangeHandler={this.filterChangeHandler}
                  />
                ) : null}
              </div>
              <div className="right">
                <nav className="vertical-list-primary flush">
                  {this.props.recentProjects ? (
                    <List.SimpleList
                      entities={this.props.recentProjects}
                      entityComponent={Project.ListItem}
                      title={"Recently Updated"}
                      icon={"manicon-bugle-small"}
                      listClasses={"flush"}
                    />
                  ) : null}
                </nav>
                <HigherOrder.Authorize entity="statistics" ability={"read"}>
                  <section>
                    <header className="section-heading-secondary">
                      <h3>
                        <i
                          className="manicon manicon-pulse-small"
                          aria-hidden="true"
                        />
                        {"Activity"}{" "}
                      </h3>
                    </header>
                    <DashboardComponents.Activity
                      statistics={this.props.statistics}
                    />
                  </section>
                </HigherOrder.Authorize>
              </div>
            </section>
          </div>
        </section>
      </div>
    );
  }
}

export default connectAndFetch(DashboardsAdminContainer);
