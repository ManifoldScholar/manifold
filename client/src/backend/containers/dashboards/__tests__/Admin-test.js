import { DashboardsAdminContainer } from "../Admin";

describe("backend/containers/dashboards/Admin", () => {
  def("project", () => factory("project", { id: 1 }));
  def("projects", () => [$project]);
  def("meta", () => ({ pagination: fixtures.pagination() }));
  def("recentProjects", () => [
    factory("project", { id: 2 }),
    factory("project", { id: 3 })
  ]);
  def("snapshotCreator", () => jest.fn());

  def("root", () => (
    <DashboardsAdminContainer
      entitiesListSearchParams={{ projects: {} }}
      projects={$projects}
      projectsMeta={$meta}
      recentProjects={$recentProjects}
      dispatch={$dispatch}
      projectsListSnapshot={{ page: 1 }}
      snapshotCreator={$snapshotCreator}
      savedSearchPaginationState={() => null}
      authentication={$state.authentication}
    />
  ));

  it("matches the snapshot when rendered", () => {
    expect(render($withApp($root))).toMatchSnapshot();
  });
});
