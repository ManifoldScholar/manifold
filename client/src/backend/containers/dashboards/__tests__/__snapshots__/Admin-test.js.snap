// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`backend/containers/dashboards/Admin matches the snapshot when rendered 1`] = `
Array [
  <div
    aria-atomic="true"
    aria-live="polite"
    class="screen-reader-text"
    role="status"
  />,
  <div>
    <h1
      class="screen-reader-text"
    >
      Dashboard
    </h1>
    <section>
      <div
        class="container"
      >
        <section
          class="backend-dashboard"
        >
          <div
            class="left"
          >
            <div
              class="dashboard-panel"
            >
              <div
                class="panel"
              >
                <div
                  class="entity-list"
                  id="entities-list-1"
                >
                  <h2
                    aria-atomic="true"
                    aria-live="polite"
                    class="entity-list__title-block entity-list__title"
                    role="status"
                  >
                    <a
                      href="/backend/projects"
                    >
                      <figure
                        class="entity-list__title-icon"
                      >
                        <svg
                          aria-hidden="true"
                          class="manicon-svg svg-icon--BEProject64"
                          fill="currentColor"
                          viewBox="0 0 64 64"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M9.103,24.948 L32.045,34.722 L54.984,24.948 L32.045,15.174 L9.103,24.948 Z M32.045,36.896 L4,24.948 L32.045,13 L60.086,24.948 L32.045,36.896 Z M60.009,32.175 L60.009,30 L32.01,41.94 L4,30.021 L4,32.194 L32.011,44.113 L60.009,32.175 Z M60.009,39.175 L60.009,37 L32.01,48.94 L4,37.02 L4,39.194 L32.011,51.113 L60.009,39.175 Z"
                            fill-rule="evenodd"
                          />
                        </svg>
                      </figure>
                      2 Projects
                    </a>
                  </h2>
                  <div
                    class="entity-list__contents-wrapper"
                  >
                    <div
                      class="entity-list__search entity-list-search"
                    >
                      <form>
                        <div
                          class="entity-list-search__keyword-row"
                        >
                          <button
                            class="entity-list-search__search-button"
                          >
                            <svg
                              aria-hidden="true"
                              class="manicon-svg svg-icon--search16"
                              fill="currentColor"
                              height="20"
                              viewBox="0 0 16 16"
                              width="20"
                              xmlns="http://www.w3.org/2000/svg"
                            >
                              <path
                                d="M7.00000003,3 C4.79086102,3 3.00000001,4.79086101 3.00000001,7.00000002 C3.00000001,9.20913903 4.79086102,11 7.00000003,11 C9.20912238,11 10.999973,9.20916605 11.0000001,7.00005001 C10.9995314,4.79106172 9.20889466,3.00044165 7.00000003,3 Z M10.8715096,10.1643028 L13.9971535,13.2899467 L13.2900467,13.9970535 L10.1644164,10.8714232 C9.30247365,11.5768009 8.20065857,12.0000001 7.00000003,12.0000001 C4.23857626,12.0000001 2,9.76142379 2,7.00000002 C2,4.23857625 4.23857626,1.99999999 7.0001,2 C9.76123985,2.00055206 11.999448,4.23876021 12.0000001,6.99990005 C12.0000001,8.2005577 11.5768322,9.30236028 10.8715096,10.1643028 Z"
                              />
                            </svg>
                            <span
                              class="screen-reader-text"
                            >
                              Search
                            </span>
                          </button>
                          <div
                            class="entity-list-search__keyword-input-wrapper"
                          >
                            <label
                              class="screen-reader-text"
                              for="1"
                            >
                              Enter Search Criteria
                            </label>
                            <input
                              class="entity-list-search__keyword-input"
                              id="1"
                              placeholder="Search..."
                              type="text"
                              value=""
                            />
                          </div>
                          <button
                            class="entity-list-search__text-button"
                          >
                            Reset
                          </button>
                          <button
                            aria-expanded="false"
                            aria-haspopup="true"
                            class="entity-list-search__text-button entity-list-search__text-button--foregrounded"
                          >
                            Options
                          </button>
                        </div>
                      </form>
                    </div>
                    <div
                      class="entity-list__header"
                    >
                      <div
                        class="entity-list__button-set"
                      />
                    </div>
                    <ul
                      aria-describedby="entities-list-instructions-1"
                      class="entity-list__list entity-list__list--rows"
                    >
                      <li
                        class="entity-row entity-list__entity"
                      >
                        <a
                          aria-describedby="1-describedby"
                          class="entity-row__row-link entity-row__row-link--block entity-row__row-link--atag"
                          href="/backend/projects/1"
                        >
                          <div
                            class="entity-row__inner entity-row__inner--in-rows entity-row__inner--with-row-link"
                          >
                            <div
                              class="entity-row__figure entity-row__figure--size-normal entity-row__figure--shape-square entity-row--figure--in-rows"
                            >
                              <svg
                                aria-hidden="false"
                                class="project-thumb-placeholder project-thumb-placeholder--all project-thumb-placeholder--primary"
                                height="100%"
                                viewBox="0 0 48 48"
                                width="100%"
                                xmlns="http://www.w3.org/2000/svg"
                              >
                                <title>
                                  Default Project Thumbnail
                                </title>
                                <g
                                  fill="none"
                                  fill-rule="evenodd"
                                >
                                  <polygon
                                    class="project-thumb-placeholder__tile"
                                    fill="#CBF7E6"
                                    points="0 42 42 42 42 0 0 0"
                                  />
                                  <g
                                    class="project-thumb-placeholder__illustration"
                                    transform="translate(9 8)"
                                  >
                                    <g
                                      transform="translate(20.683 2.244)"
                                    >
                                      <polygon
                                        points=".161 2.954 .161 21.013 1.612 23.598 3.06 21.013 3.06 2.954"
                                      />
                                      <path
                                        d="M3.06081501,21.012797 L0.161633021,21.012797"
                                      />
                                      <path
                                        d="M0.162011257,2.95560225 L3.06119325,2.95560225 L3.06119325,1.53299812 C3.06119325,0.980713374 2.613478,0.532998124 2.06119325,0.532998124 L1.16201126,0.532998124 C0.609726507,0.532998124 0.162011257,0.980713374 0.162011257,1.53299812 L0.162011257,2.95560225 Z"
                                        stroke-linejoin="round"
                                      />
                                    </g>
                                    <g
                                      transform="translate(.195 11.463)"
                                    >
                                      <polygon
                                        points="14.898 .419 .985 .419 .985 11.282 8.141 11.282 8.141 14.378 10.844 11.282 14.898 11.282"
                                      />
                                      <path
                                        d="M12.5237854 3.46905816L3.32696735 3.46905816M12.5237854 8.00165403L3.32696735 8.00165403M12.5237854 5.73507242L3.32696735 5.73507242"
                                      />
                                    </g>
                                    <g
                                      transform="translate(.195 .195)"
                                    >
                                      <path
                                        d="M11.8853223 5.12585966C11.8853223 6.3608015 10.8848871 7.36123677 9.64994522 7.36123677 8.41689456 7.36123677 7.41645929 6.3608015 7.41645929 5.12585966 7.41645929 3.89091782 8.41689456 2.89048255 9.64994522 2.89048255 10.8848871 2.89048255 11.8853223 3.89091782 11.8853223 5.12585966zM11.8853223 5.23573734L15.0417051 2.0793546C15.6109508 1.51010882 16.5849096 1.91293058 16.5849096 2.71857411M3.99814784 3.09718874L6.63067317.464663415C7.20181013-.106473546 8.17576886.2982394 8.17576886 1.10388293"
                                      />
                                      <path
                                        d="M5.22155347 5.12585966C5.22155347 6.3608015 4.2211182 7.36123677 2.98617636 7.36123677 1.7531257 7.36123677.752690432 6.3608015.752690432 5.12585966.752690432 3.89091782 1.7531257 2.89048255 2.98617636 2.89048255 4.2211182 2.89048255 5.22155347 3.89091782 5.22155347 5.12585966zM5.22155347 5.34334559L5.71136961 5.13720675C6.09906191 4.97267392 6.53781614 4.97267392 6.92739962 5.13720675L7.41532458 5.34334559"
                                      />
                                    </g>
                                  </g>
                                  <g
                                    class="project-thumb-placeholder__frame"
                                    stroke="#828282"
                                  >
                                    <polyline
                                      points="48 6 48 48 6 48"
                                    />
                                    <polyline
                                      points="45 3 45 45 3 45"
                                    />
                                    <polygon
                                      points="0 42 42 42 42 0 0 0"
                                    />
                                  </g>
                                </g>
                              </svg>
                            </div>
                            <div
                              class="entity-row__text entity-row__text--in-rows"
                            >
                              <h3
                                class="entity-row__title entity-row__title--in-rows"
                              >
                                <span
                                  class="entity-row__title-inner"
                                >
                                  <span>
                                    Rowan Test
                                  </span>
                                </span>
                                <span
                                  class="aria-describedby"
                                  id="1-describedby"
                                >
                                  View Rowan Test
                                </span>
                              </h3>
                              <h4
                                class="entity-row__subtitle entity-row__subtitle--in-rows"
                              >
                                <span />
                              </h4>
                              <div
                                class="entity-row__meta entity-row__meta--in-rows"
                              />
                            </div>
                          </div>
                        </a>
                      </li>
                    </ul>
                    <div
                      class="entity-list__pagination"
                    >
                      <nav
                        aria-label="Pagination"
                        class="list-pagination"
                      >
                        <ul
                          class="list-pagination__columns"
                        >
                          <li
                            class="list-pagination__column"
                          >
                            <ul>
                              <li
                                class="list-pagination__page list-pagination__page--prev list-pagination__page--disabled"
                              >
                                <a
                                  disabled=""
                                  href="#entities-list-1"
                                >
                                  <svg
                                    aria-hidden="true"
                                    class="manicon-svg svg-icon--arrowLongLeft16"
                                    fill="currentColor"
                                    viewBox="0 0 24 16"
                                    xmlns="http://www.w3.org/2000/svg"
                                  >
                                    <path
                                      d="M19.9210555,8.50098377 L2,8.50098377 L2,7.49848233 L19.9183709,7.49848233 L14.6653905,2.76142495 L15.3486096,2 L22.0012062,7.99920999 L15.3651285,14 L14.6808716,13.2395122 L19.9210555,8.50098377 Z"
                                      transform="rotate(180 12 8)"
                                    />
                                  </svg>
                                  <span
                                    class="list-pagination__icon-label"
                                  >
                                    PREV
                                  </span>
                                </a>
                              </li>
                            </ul>
                          </li>
                          <li
                            class="list-pagination__column-middle"
                          >
                            <ul
                              class="list-pagination__pages"
                            >
                              <li
                                class="list-pagination__page list-pagination__page--ordinal list-pagination__page--disabled"
                              >
                                <a
                                  href="#entities-list-1"
                                >
                                  <span
                                    aria-hidden="true"
                                  >
                                    1
                                  </span>
                                  <span
                                    class="screen-reader-text"
                                  >
                                    Go to page: 1
                                  </span>
                                </a>
                              </li>
                              <li
                                class="list-pagination__page list-pagination__page--ordinal"
                              >
                                <a
                                  href="#entities-list-1"
                                >
                                  <span
                                    aria-hidden="true"
                                  >
                                    2
                                  </span>
                                  <span
                                    class="screen-reader-text"
                                  >
                                    Go to page: 2
                                  </span>
                                </a>
                              </li>
                            </ul>
                          </li>
                          <li
                            class="list-pagination__column"
                          >
                            <ul>
                              <li
                                class="list-pagination__page list-pagination__page--next"
                              >
                                <a
                                  href="#entities-list-1"
                                >
                                  <span
                                    class="list-pagination__icon-label"
                                  >
                                    NEXT
                                  </span>
                                  <svg
                                    aria-hidden="true"
                                    class="manicon-svg svg-icon--arrowLongRight16"
                                    fill="currentColor"
                                    viewBox="0 0 24 16"
                                    xmlns="http://www.w3.org/2000/svg"
                                  >
                                    <path
                                      d="M19.9210555,8.50098377 L2,8.50098377 L2,7.49848233 L19.9183709,7.49848233 L14.6653905,2.76142495 L15.3486096,2 L22.0012062,7.99920999 L15.3651285,14 L14.6808716,13.2395122 L19.9210555,8.50098377 Z"
                                    />
                                  </svg>
                                </a>
                              </li>
                            </ul>
                          </li>
                        </ul>
                      </nav>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="right"
          >
            <div
              class="dashboard-panel"
            >
              <div
                class="panel"
              >
                <div
                  class="entity-list"
                  id="entities-list-1"
                >
                  <h2
                    class="entity-list__title-block entity-list__title"
                  >
                    <figure
                      class="entity-list__title-icon"
                    >
                      <svg
                        aria-hidden="true"
                        class="manicon-svg svg-icon--BEProject64"
                        fill="currentColor"
                        viewBox="0 0 64 64"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M9.103,24.948 L32.045,34.722 L54.984,24.948 L32.045,15.174 L9.103,24.948 Z M32.045,36.896 L4,24.948 L32.045,13 L60.086,24.948 L32.045,36.896 Z M60.009,32.175 L60.009,30 L32.01,41.94 L4,30.021 L4,32.194 L32.011,44.113 L60.009,32.175 Z M60.009,39.175 L60.009,37 L32.01,48.94 L4,37.02 L4,39.194 L32.011,51.113 L60.009,39.175 Z"
                          fill-rule="evenodd"
                        />
                      </svg>
                    </figure>
                    Recently Updated
                  </h2>
                  <div
                    class="entity-list__contents-wrapper"
                  >
                    <ul
                      aria-describedby="entities-list-instructions-1"
                      class="entity-list__list entity-list__list--rows"
                    >
                      <li
                        class="entity-row entity-list__entity"
                      >
                        <a
                          aria-describedby="1-describedby"
                          class="entity-row__row-link entity-row__row-link--block entity-row__row-link--atag"
                          href="/backend/projects/2"
                        >
                          <div
                            class="entity-row__inner entity-row__inner--in-rows entity-row__inner--with-row-link"
                          >
                            <div
                              class="entity-row__text entity-row__text--in-rows"
                            >
                              <h3
                                class="entity-row__title entity-row__title--in-rows"
                              >
                                <span
                                  class="entity-row__title-inner"
                                >
                                  <span>
                                    Rowan Test
                                  </span>
                                </span>
                                <span
                                  class="aria-describedby"
                                  id="1-describedby"
                                >
                                  View Rowan Test
                                </span>
                              </h3>
                              <h4
                                class="entity-row__subtitle entity-row__subtitle--in-rows"
                              >
                                <span />
                              </h4>
                              <div
                                class="entity-row__meta entity-row__meta--in-rows"
                              />
                            </div>
                          </div>
                        </a>
                      </li>
                      <li
                        class="entity-row entity-list__entity"
                      >
                        <a
                          aria-describedby="1-describedby"
                          class="entity-row__row-link entity-row__row-link--block entity-row__row-link--atag"
                          href="/backend/projects/3"
                        >
                          <div
                            class="entity-row__inner entity-row__inner--in-rows entity-row__inner--with-row-link"
                          >
                            <div
                              class="entity-row__text entity-row__text--in-rows"
                            >
                              <h3
                                class="entity-row__title entity-row__title--in-rows"
                              >
                                <span
                                  class="entity-row__title-inner"
                                >
                                  <span>
                                    Rowan Test
                                  </span>
                                </span>
                                <span
                                  class="aria-describedby"
                                  id="1-describedby"
                                >
                                  View Rowan Test
                                </span>
                              </h3>
                              <h4
                                class="entity-row__subtitle entity-row__subtitle--in-rows"
                              >
                                <span />
                              </h4>
                              <div
                                class="entity-row__meta entity-row__meta--in-rows"
                              />
                            </div>
                          </div>
                        </a>
                      </li>
                    </ul>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </section>
      </div>
    </section>
  </div>,
]
`;
